crates/primitives/src/transaction/mod.rs

```diff
@@ -47,6 +47,100 @@ pub(crate) static PARALLEL_SENDER_RECOVERY_THRESHOLD: Lazy<usize> =
         _ => 5,
     });
 
+/// Value is the type of the `value` field in the various Ethereum transactions structs.
+/// While the field is 256 bits, for many chains it's not possible for the field to use
+/// this full precision, hence we use a wrapper type to allow for overriding of encoding.
+#[derive_arbitrary(compact, rlp)]
+#[derive(Default, Debug, Copy, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Value(U256);
+
+impl From<Value> for U256 {
+    /// unwrap Value to U256
+    fn from(value: Value) -> U256 {
+        value.0
+    }
+}
+
+impl<T> From<T> for Value
+where
+    Self: UintTryFrom<T>,
+{
+    /// construct a Value from misc. other uint types
+    fn from(value: T) -> Self {
+        match Self::uint_try_from(value) {
+            Ok(n) => n,
+            Err(e) => panic!("Uint conversion error: {e}"),
+        }
+    }
+}
+
+impl UintTryFrom<U256> for Value {
+    #[inline]
+    fn uint_try_from(value: U256) -> Result<Self, ToUintError<Self>> {
+        Ok(Self(value))
+    }
+}
+
+impl UintTryFrom<u128> for Value {
+    #[inline]
+    fn uint_try_from(value: u128) -> Result<Self, ToUintError<Self>> {
+        Ok(Self(U256::from(value)))
+    }
+}
+
+impl UintTryFrom<u64> for Value {
+    #[inline]
+    fn uint_try_from(value: u64) -> Result<Self, ToUintError<Self>> {
+        Ok(Self(U256::from(value)))
+    }
+}
+
+impl Encodable for Value {
+    fn encode(&self, out: &mut dyn bytes::BufMut) {
+        self.0.encode(out)
+    }
+    fn length(&self) -> usize {
+        self.0.length()
+    }
+}
+
+impl Decodable for Value {
+    fn decode(buf: &mut &[u8]) -> Result<Self, DecodeError> {
+        Ok(Value(U256::decode(buf)?))
+    }
+}
+
+/// As ethereum circulation is around 120mil eth as of 2022 that is around
+/// 120000000000000000000000000 wei we are safe to use u128 for Value's encoding
+/// as its max number is 340282366920938463463374607431768211455.
+/// This optimization should be disabled for chains such as Optimism, where
+/// some tx values may require full 256-bit precision.
+impl Compact for Value {
+    #[allow(unreachable_code)]
+    fn to_compact<B>(self, buf: &mut B) -> usize
+    where
+        B: bytes::BufMut + AsMut<[u8]>,
+    {
+        #[cfg(test)]
+        {
+            return self.0.to_compact(buf)
+        }
+        let i: u128 = self.0.uint_try_to().expect("value could not be converted to u128");
+        i.to_compact(buf)
+    }
+
+    #[allow(unreachable_code)]
+    fn from_compact(buf: &[u8], identifier: usize) -> (Self, &[u8]) {
+        #[cfg(test)]
+        {
+            let (i, buf) = U256::from_compact(buf, identifier);
+            return (Value(i), buf)
+        }
+        let (i, buf) = u128::from_compact(buf, identifier);
+        (Value::from(i), buf)
```

Comment:
why the different `cfg(test)`?

Reply:
Because otherwise the fuzz tests fail when they try to encode a full precision U256.   Open to suggestions on how to better handle this.